llvm/lib/Transforms/Vectorize/VPlan.h

```diff
@@ -1405,11 +1407,16 @@ class VPInstruction : public VPRecipeWithIRFlags {
 class VPWidenRecipe : public VPRecipeWithIRFlags {
   unsigned Opcode;
 
+protected:
+  template <typename IterT>
+  VPWidenRecipe(unsigned VPDefOpcode, Instruction &I,
+                iterator_range<IterT> Operands)
+      : VPRecipeWithIRFlags(VPDefOpcode, Operands, I), Opcode(I.getOpcode()) {}
+
 public:
   template <typename IterT>
   VPWidenRecipe(Instruction &I, iterator_range<IterT> Operands)
-      : VPRecipeWithIRFlags(VPDef::VPWidenSC, Operands, I),
-        Opcode(I.getOpcode()) {}
+      : VPWidenRecipe(VPDef::VPWidenSC, I, Operands) {}
```

Comment:
This should probably have a custom `classof` is `isa<VPWidenRecipe>` returns true for both VPWidenRecipe and VPWidenEVLRecipe?

Reply:
I assume than you really meant to introduce base class for `VPWidenRecipe` and `VPWidenEVLRecipe` that will return true for both of them, right ? In this case class hierarchy will be similar to `VPWiden[Load|Store][|EVL]`.
If so, same should go to all future EVL-recipes:
```
           VPSomeRecipeBase
           /             \
VPSomeRecipe          VPSomeEVLRecipe
```

```
VPSomeRecipeBase::classof(...) { return Opcode == VPSomeRecipeSC || Opcode == VPSomeEVLRecipeSC; }
```

Reply:
Ideally VPWidenRecipe could serve as such a base class as mentioned above, unless there’s a compelling reason not to. That way we automatically benefit from all folds and analysis already implemented for VPWidenRecipe

Reply:
Not sure I understand added hierarchy then. If the goal is to allow reuse of existing code, then what is different in `VPWidenStoreEVL` or `VPWidenLoadEVL` as they won't be treated as `VPWidenStore` or `VPWidenLoad` respectively: 
```
isa<VPWidenMemoryRecipe>(EVLStore) => true
isa<VPWidenStoreEVLRecipe>(EVLStore) => true
isa<VPWidenStoreRecipe>(EVLStore) => false
```
Anyway, I'm _ok_ to extend classof for now

Reply:
Thanks for the update. Adjusting `VPWidenStoreEVLRecipe`/`VPWidenLoadEVLRecipe` would probably make sense separately, although the reasoning may be a bit more complicated.